src/vs/workbench/contrib/notebook/browser/contrib/multicursor/notebookMulticursor.ts

```diff
@@ -258,10 +264,67 @@ export class NotebookMultiCursorController extends Disposable implements INotebo
 		}));
 	}
 
+	private updateFinalUndoRedo() {
+		const anchorCellModel = this.anchorCell?.[1].getModel();
+		if (!anchorCellModel) {
+			// should not happen
+			return;
+		}
+
+		const newElementsMap: ResourceMap<IUndoRedoElement[]> = new ResourceMap<IUndoRedoElement[]>();
+		const resources: URI[] = [];
+
+		this.trackedMatches.forEach(trackedMatch => {
+			const undoRedoState = trackedMatch.undoRedoHistory;
+			if (!undoRedoState) {
+				return;
+			}
+
+			resources.push(trackedMatch.cellViewModel.uri);
+
+			const currentPastElements = this.undoRedoService.getElements(trackedMatch.cellViewModel.uri).past.slice();
+			const oldPastElements = trackedMatch.undoRedoHistory.past.slice();
+			const newElements = currentPastElements.slice(oldPastElements.length);
+			if (newElements.length === 0) {
+				return;
+			}
+
+			newElementsMap.set(trackedMatch.cellViewModel.uri, newElements);
+
+			this.undoRedoService.removeElements(trackedMatch.cellViewModel.uri);
+			oldPastElements.forEach(element => {
+				this.undoRedoService.pushElement(element);
+			});
+		});
+
+		this.undoRedoService.pushElement({
+			type: UndoRedoElementType.Workspace,
+			resources: resources,
+			label: 'Multi Cursor Edit',
+			code: 'multiCursorEdit',
+			confirmBeforeUndo: false,
+			undo: async () => {
+				newElementsMap.forEach(async value => {
+					value.reverse().forEach(async element => {
+						await element.undo();
+					});
```

Comment:
These are async, but do they need to happen in series to get executed correctly? Atm this runs all of them in parallel

Reply:
@connor4312 they don't need to since they are separate models. Though I would later make them a "transaction" to avoid issues with multiple undo/redo requests.